Enhancement to CDK/Linux/hello-world example

hu-jin-aws opened the pull request:
This PR is an enhancement to the code under `CDK/Linux/hello-world` to support user specifying AWS Management Components in the EC2 Image Builder Pipeline configurations. It also included some general cleanup. Below are the detailed list of the changes:

- Update formatting and misc cosmetics fix
- Support multiple components in one recipe
- Support AWS managed components
- Support specifying path to single spec file
- Remove unused S3 bucket
- A few other misc fixes
- Update README

austoonz commented on CDK/Linux/hello-world/README.md:
```diff
@@ -16,10 +16,10 @@ The following settings can be configured before running CDK deployment. Those se
 | ImageBuilderPipelineConfigurations | List | A list of configuration settings to define the EC2 Image Building pipelines. Each entry in the list defines an Image Building Pipeline (See next Section for more information about this key) | Default settings for a sample pipeline | Yes      |
 
-* notice `AlejandroRosalez@example.com` is a fictious email address for demo purpose, you can replace it with your own email addresses.
+- notice `AlejandroRosalez@example.com` is a fictious email address for demo purpose, you can replace it with your own email addresses.
```
Not too fussed, but here's no need to change the `*` to a `-`, they're [both supported](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#lists) by GitHub for creating unordered lists.

hu-jin-aws replied:
Thanks, that was done by the VS Code auto-formatting, I can change it back if you prefer.

hu-jin-aws replied:
udpated.

austoonz commented on CDK/Linux/hello-world/README.md:
```diff
@@ -110,7 +112,14 @@ Below are an example of the value in `ImageBuilderPipelineConfigurations` key th
+
+### Output of the Pipeline
+
+After an EC2 Image Builder image pipeline is established after deploying this stack with proper configuration, it will start automatically to build the EC2 AMI image based on the components specified in the configuration. The pipeline will take a while to complete and when it is completed, the ID of the new AMI image that is built can be view from the following two locations:
```
1. Can you please replace "EC2 AMI image" and with simply "AMI"?
2. Can you replace "the ID of the new AMI image that is built can be view from the following two locations" with "the ID of the new AMI that is built can be found in the following locations:"

hu-jin-aws replied:
Thanks. Fixed as you suggested

austoonz commented on CDK/Linux/hello-world/README.md:
```diff
@@ -110,7 +112,14 @@ Below are an example of the value in `ImageBuilderPipelineConfigurations` key th
+
+After an EC2 Image Builder image pipeline is established after deploying this stack with proper configuration, it will start automatically to build the EC2 AMI image based on the components specified in the configuration. The pipeline will take a while to complete and when it is completed, the ID of the new AMI image that is built can be view from the following two locations:
+
+- In `Output images` section when open EC2 Image Builder AWS Console page and click `Image pipelines` on the left navigation bar and then click the pipeline that is created by this stack.
```
```
- In the EC2 Image Builder console, select `Image pipelines` from the left navigation bar, then select the pipeline created by this stack, then look under `Output images`.
```

hu-jin-aws replied:
Fixed as you suggested

austoonz commented on CDK/Linux/hello-world/README.md:
```diff
@@ -110,7 +112,14 @@ Below are an example of the value in `ImageBuilderPipelineConfigurations` key th
+After an EC2 Image Builder image pipeline is established after deploying this stack with proper configuration, it will start automatically to build the EC2 AMI image based on the components specified in the configuration. The pipeline will take a while to complete and when it is completed, the ID of the new AMI image that is built can be view from the following two locations:
+
+- In `Output images` section when open EC2 Image Builder AWS Console page and click `Image pipelines` on the left navigation bar and then click the pipeline that is created by this stack.
+- A parameter is created in AWS System Manager Parameter Store when deploying the pipeline and named as `imagebuilder_ami_<pipeline name>`. The id of the new AMI built by the pipeline will be stored in this parameter so that it can be read by other stacks if necessary. Before the pipeline complete building the AMI image, the value of this parameter is `n/a`.
```
1. "The id of the new AMI" -> "The ID of the new AMI"
2. "Before the pipeline complete building the AMI image, the value" -> "Before the pipeline finishes building the AMI, the value"

hu-jin-aws replied:
Fixed as you suggested

austoonz commented on CDK/Linux/hello-world/cdk.json:
```diff
@@ -4,21 +4,20 @@
+        "cfnImageRecipeName": "imagebuilder-example-recipe",
+        "version": "1.0.0",
+        "parentImage": {
+          "ap-southeast-2": { "amiID": "ami-0b7dcd6e6fd797935" },
+          "ap-southeast-1": { "amiID": "ami-055d15d9cfddf7bd3" },
+          "us-east-1": { "amiID": "ami-04505e74c0741db8d" },
+          "us-east-2": { "amiID": "ami-0fb653ca2d3203ac1" },
+          "us-west-1": { "amiID": "ami-01f87c43e618bf8f0" },
+          "us-west-2": { "amiID": "ami-0892d3c7ee96c0bf7" }
```
Are we able to update these to the latest Ubuntu 20.04 AMIs?

Eg, the one for `ap-southeast-2` was created on 30 November 2021, whereas the latest (`ami-02e8cbf7681c3ae51`) is from 10 August 2022.

---

An improvement here would be to list the EC2 Image Builder ARN for the parent image instead of the Image ID's directly, eg `arn:aws:imagebuilder:us-east-1:aws:image/ubuntu-server-20-lts-arm64/x.x.x` (for us-east-1)

hu-jin-aws replied:
Those AMI IDs has been updated to the image builder ARNs for latest Ubuntu Server 20 LTS x86. The code can accommodate both image builder arn and ami ID in this field. Also the configuration in this file is only an example. When user use this sample to start build their AMI images, they should check and update those parent image IDs to make sure the base image is what they will need.

austoonz commented on CDK/Linux/hello-world/lib/aws-image-builder-stack.ts:
```diff
@@ -75,62 +50,114 @@ export class ImageBuilderStack extends Stack {
+    for (const component of components) {
+      // test if component is a valid AWS managed component arn
+      const arnMatch = component.match(
+        /^arn:aws:imagebuilder:[a-z0-9\-]*:aws:component\/([\w\-]*)\/.*/
```
Given this sample could be tried in any partition, can you adjust to not use the `aws` partition directly in this regex?

hu-jin-aws replied:
Regex is updated to accommodate all known partition values at the moment.

austoonz reviewed (changes requested):
Primary some text formatting in the README.

Regarding the code request to update to latest Image ID, or better to use the Image Builder ARN, I'm happy for simply updating to the latest Image ID, but if so a future PR to change this to the Image Builder ARNs would be preferable, so any customers basing work off this sample don't launch potentially year+ old AMIs.

austoonz commented on CDK/Linux/hello-world/lib/aws-image-builder-stack.ts:
```diff
@@ -97,7 +97,7 @@ export class ImageBuilderStack extends Stack {
       // test if component is a valid AWS managed component arn
       const arnMatch = component.match(
-        /^arn:aws:imagebuilder:[a-z0-9\-]*:aws:component\/([\w\-]*)\/.*/
+        /^arn:aws:imagebuilder:[a-z0-9\-]*:aws-?[a-z0-9\-]*:component\/([\w\-]*)\/.*/
```
Unfortunately this change is on the wrong `aws` string.

It needs to be on the one after `arn`, you could use: `arn:(?:aws|aws-cn|aws-us-gov|aws-iso|aws-iso-b):imagebuilder:......`

hu-jin-aws replied:
Oops. Thanks for spotting this mistake. I have fixed the regex. I was trying to use the wildcard to fit more possible future combination of the partition values, but I have updated to use yours. I think we can always update the code when there is any new partition.

austoonz reviewed (changes requested):
Two minor changes. One a simply typo, the other was the regex change to the wrong `aws` statement.

hu-jin-aws commented on CDK/Linux/hello-world/README.md:
```diff
@@ -116,10 +116,10 @@ The new component script must be provided before the new image building pipeline
-- In `Output images` section when open EC2 Image Builder AWS Console page and click `Image pipelines` on the left navigation bar and then click the pipeline that is created by this stack.
-- A parameter is created in AWS System Manager Parameter Store when deploying the pipeline and named as `imagebuilder_ami_<pipeline name>`. The id of the new AMI built by the pipeline will be stored in this parameter so that it can be read by other stacks if necessary. Before the pipeline complete building the AMI image, the value of this parameter is `n/a`.
+- In the EC2 Image Builder console, select `Image pipelines` from the left navigation bar, then select the pipeline created by this stack, then look under `Output images`.
+- A parameter is created in AWS System Manager Parameter Store when deploying the pipeline and named as `imagebuilder_ami_<pipeline name>`. The ID of the new AMI built by the pipeline will be stored in this parameter so that it can be read by other stacks if necessary. Before the pipeline finishs building the AMI, the value of this parameter is `n/a`.
```
Thanks! It is fixed.

hu-jin-aws commented:
@austoonz To keep the history clean I have rebased two new commits of addressing review comments into the original commit, therefore it might be a little hard to see the changes when you review. Apologies for that.
